New landing page design

simonsteiger opened the pull request:
Related to #101.

This PR contains a draft for a new landing page and the related styling changes.

Still missing in this draft are:

- improved footer
- slideshow feature for news
- information as listed in #76

I'm happy to continue iterating on this, but thought that it might be a good time to discuss what I've currently got.

shravanngoswamii commented:
Just a suggestion, IMO text color of nav-bar in dark mode is not quite visible properly, it should have more light, and normal text is too much bright for dark mode... maybe it's just my eyesight delusion😅! 
![image](https://github.com/user-attachments/assets/aa733a57-4f3c-4ef0-bba4-395f90dca778)
![image](https://github.com/user-attachments/assets/ba4d306a-d7fb-4c8f-9839-68359610b3b5)

shravanngoswamii commented:
@simonsteiger Can we try to add something like this SVG, we created this rough sketch, it's a bit unclean, but I think it worth a shot!

Adding @Govindggupta to this conversation as he created this SVG a while back!

https://github.com/user-attachments/assets/fe0b810f-525a-43e3-9999-8f3e55d021f8

![image](https://github.com/user-attachments/assets/43afed1f-ea86-4494-9988-ecc833967df7)

<details><summary>Code</summary>
<p>

```html
<!DOCTYPE html>
<html lang="en">
   <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>SVG Animation</title>
      <style>
         body {
         margin: 0;
         padding: 0;
         width: 100%;
         height: 100%;
         }
         .line-graph {
         /* height: 85vh; */
         width: 100%;
         padding-top: 5%;
         position: relative;
         }
         svg {
         width: 100%;
         height: 100%;
         }
         @keyframes draw {
         to {
         stroke-dashoffset: 0;
         }
         }
         .line {
         fill: none;
         stroke-width: 7;
         stroke-dasharray: 3900;
         stroke-dashoffset: 3900;
         animation: draw 6s forwards ease-in-out;
         stroke-opacity: 1;
         }
         .line1 {
         stroke: #b352cc;
         animation-delay: 0s;
         }
         .line2 {
         stroke: #cc3333;
         animation-delay: 0.5s;
         }
         .line3 {
         stroke: #218921;
         animation-delay: 1s;
         }
         .upperContent {
         width: 90%;
         text-align: end;
         padding-right: 10%;
         position: absolute;
         z-index: 5;
         opacity: 0;
         animation: fadeInText 2s forwards ease-in-out;
         animation-delay: 3s;
         }
         .upperContent>h1 {
         font-size: 80px;
         font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
         }
         .upperContent>p {
         font-size: 25px;
         font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
         margin-top: 5%;
         }
         .upperContent>h1,
         .upperContent>p {
         margin: 0;
         padding: 0;
         }
         @keyframes fadeInText {
         to {
         opacity: 1;
         transform: translateY(0);
         }
         }
         nav {
         width: 100%;
         height: 60px;
         background-color: #073c44;
         text-align: center;
         color: whitesmoke;
         font-size: 50px;
         }
      </style>
   </head>
   <body>
      <nav>
         NAVBAR
      </nav>
      <div class="line-graph">
         <div class="upperContent">
            <h1>Turing.jl</h1>
            <p>Bayesian inference with probabilistic programming</p>
         </div>
         <div class="Animation">
            <svg width="2300" height="498" viewBox="0 0 2300 498" fill="none" xmlns="http://www.w3.org/2000/svg">
               <path class="line line1"
                  d="M0 487C428.737 486.5 424.61 478.535 493.575 389C523.23 350.5 595.105 213 654.414 213C713.724 213 748.907 290 790.625 344.5C832.343 399 846.416 425.5 908.239 461.5C979.498 502.995 1144.47 487 2300 487" />
               <path class="line line2"
                  d="M0 485.5C212.654 485.5 378.977 486 565.45 486C751.923 486 763.644 514.469 852.448 292.5C881.709 219.36 911.829 96 950.052 96C988.276 96 1016.27 197.222 1050.19 287.031C1116.15 461.682 1136.64 484.016 1322.06 484.016C1416.28 484.016 2076.04 486.5 2300 486.5" />
               <path class="line line3"
                  d="M0 486.499C101.007 486.499 207.547 486.499 308.554 486.499C409.56 486.499 483.491 486.499 577.149 486.499C668.492 486.499 699.329 489.286 731.197 460.973C773.165 423.687 790.94 330.837 809.112 262.111C827.285 193.386 858.089 8 890.182 8C922.275 8 944 162.661 971.155 262.111C988.985 327.407 1006.43 426.95 1058.05 469.982C1081.26 489.327 1164.7 486.499 1248.64 486.499C1356.96 486.499 1406.25 486.499 1516.74 486.499C1627.23 486.499 2191.68 487 2300 487" />
            </svg>
         </div>
      </div>
   </body>
</html>
```

</p>
</details>

simonsteiger commented:
> Just a suggestion, IMO text color of nav-bar in dark mode is not quite visible properly, it should have more light, and normal text is too much bright for dark mode... maybe it's just my eyesight delusion😅!

Thanks for the feedback, those are good points. I'll have a look at the colours again.

The SVG looks awesome! I'll try to integrate it into the current design in the coming days. I might also experiment with moving the "Turing.jl" header and subheader around a bit to align it better with the current content layout.

shravanngoswamii commented:
> The SVG looks awesome! I'll try to integrate it into the current design in the coming days. I might also experiment with moving the "Turing.jl" header and subheader around a bit to align it better with the current content layout.

Feel free to experiment anything! Let us know if you need any help!

simonsteiger commented:
I tried to make the SVG header work as well as possible on mobile screens.

Do you have other ideas for the layout on mobile screens? I played with a few versions but preferred one where "Turing.jl" is the most prominent (instead of smaller and to the right of the distributions):

<img width="436" alt="image" src="https://github.com/user-attachments/assets/2442ec95-3c07-4d4b-88d8-3393367c325a">

shravanngoswamii commented:
> I tried to make the SVG header work as well as possible on mobile screens.

Thank you, @simonsteiger!

> Do you have other ideas for the layout on mobile screens? I played with a few versions but preferred one where "Turing.jl" is the most prominent (instead of smaller and to the right of the distributions)

It's looking good, I will let you if I get any other ideas!

shravanngoswamii commented:
@yebai @penelopeysm Please let us know if you have any suggestions to complete this PR!

penelopeysm commented:
Hello @simonsteiger! I think the layout is very beautiful and elegant, and having done a bit of web design myself I know that tweaking CSS to make something look sleek is not trivial so I wanted to say that this is really, really good work :) Thank you!

Some small comments:

- In the SVG, it would be good if the text 'Turing.jl' would appear earlier. Right now you have to wait ~ 3 seconds before it appears.
   Would it be possible to move the text to the left-hand side and the peaks of the Gaussians to the right-hand side?

- I like the libraries page a lot, but I don't think it deserves to be put front and centre together with tutorials and the link to the github organisation. The average user wouldn't really be navigating there very often imo.
   - In its place, perhaps we could have a link to #turing on Julia Slack (https://julialang.slack.com/archives/CCYDC34A0) and/or the probabilistic programming forum on Discourse (https://discourse.julialang.org/c/domain/probprog/48). These are places where people can get in touch with us / other users. I think that would be more welcoming and pertinent

- Is there a reason why the tutorial button is a different colour from the others? If there isn't a very important reason, I would prefer to have it be the same as UI consistency is really important.

I'm happy to continue discussing what other, additional, information is worth including. However, I _strongly_ believe that that should be part of a separate PR, not this one.

Unrelated to this PR. but my moonshot dream for the front page is to have a 'playground' - people can edit a model definition, select a sampler from a dropdown, then click a button and have it sample and display plots. I know this is not really possible (this would require either Julia -> WASM compilation, or a dedicated server machine) but just putting it out there. 😄

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,190 @@
+<div class="list grid quarto-listing-cols-3">
+  <a class="pseudolisting" href="https://discourse.julialang.org/c/domain/probprog/48">
+    <div class="card-title h5">
+      Discourse
+    </div>
+    <p class="quarto-grid-item">
+      Get support from and discuss with other users.
+    </p>
+  </a>
+  <a class="pseudolisting" href="https://turinglang.org/docs/tutorials/docs-00-getting-started/">
+    <div class="card-title h5">
+      Documentation
+    </div>
+    <p class="quarto-grid-item">
+      Familiarize yourself with the Turing.jl API.
+    </p>
+  </a>
+  <a class="pseudolisting" href="https://turinglang.org/docs/tutorials/00-introduction/">
+    <div class="card-title h5">
+      Official tutorials
+    </div>
+    <p class="quarto-grid-item">
+      Learn to tackle specific problems with Turing.jl.
+    </p>
+  </a>
+  <a class="pseudolisting" href="resources/content/inofficial_tutorials.qmd">
+    <div class="card-title h5">
+      Inofficial tutorials
+    </div>
+    <p class="quarto-grid-item">
+      A list of inofficial tutorials using Turing.jl.
+    </p>
+  </a>
+  <a class="pseudolisting" href="resources/content/talks.qmd">
+    <div class="card-title h5">
+      Talks
+    </div>
+    <p class="quarto-grid-item">
+      A list of talks on the Turing.jl ecosystem.
+    </p>
```
This is the sort of code I'm talking about – it's repetitive, and that's just the nature of raw HTML. In a framework like Astro you would be able to declare something like `<GridItem link="..." title="...">`.

simonsteiger replied:
+1 for Astro! It would be a great choice if it possible to integrate the existing docs workflow.

While working on this PR, I was simultaneously working on another "docs-like" page that uses [Astro Starlight](https://starlight.astro.build/). I often felt bad repeating all these HTML chunks over and over and wished I could just wrap them in a component...

I don't think I'm a good web dev, but I think it's a fun creative outlet. If you ever go through with the move to Astro or another proper JS framework, I'd be happy to help in the effort.

penelopeysm replied:
I probably wasn't clear enough in my original comment, sorry. I think we can still build the docs repo with Quarto and this repo with Astro (or similar) It sounds like a good follow-up stretch goal 😄 I'll open an issue to track that.

penelopeysm replied:
https://github.com/TuringLang/turinglang.github.io/issues/110

simonsteiger replied:
Ah, sorry, I think your explanation was clear! I phrased the "integrate the existing docs workflow" poorly. That should have read "integrate with the existing quarto docs workflow", oopsies.

penelopeysm reviewed (commented):
Finally, given the amount of custom HTML code that is in this, I wonder whether it is still worth sticking with Quarto?

In my opinion, something like [Astro](https://astro.build/) would be better as this will still allow you to have page content in Markdown files, but it also allows you to build components and therefore make the HTML code more composable and reusable. Of course we could go the full distance and use a proper JS framework like React or (my favourite) Svelte.

It also makes it easier to add new features in the future (not just new content) as we'd be using tools meant for web design instead of trying to shoehorn web design into Quarto.

Quarto is pretty much irreplaceable for the docs repo because that involves running Julia code in notebooks, but I don't think we have any code execution in this repo.

As per my previous comment, I don't really think that this is something that should be changed now, but IMO it's something worth thinking about going forward, and I probably wouldn't mind taking that on.

shravanngoswamii commented:
> Unrelated to this PR. but my moonshot dream for the front page is to have a 'playground' - people can edit a model definition, select a sampler from a dropdown, then click a button and have it sample and display plots. I know this is not really possible (this would require either Julia -> WASM compilation, or a dedicated server machine) but just putting it out there. 😄

Maybe these can help, just putting them here if we take this on in future!
https://tshort.github.io/WebAssemblyCompiler.jl/stable/
https://dspace.mit.edu/handle/1721.1/150151
https://discourse.julialang.org/t/julia-and-wasm/108482

simonsteiger commented:
Hi @penelopeysm! I was really happy to read your comment – working on the design was a lot of fun, and it's of course great to hear that you think it looks beautiful and elegant.

> * In the SVG, it would be good if the text 'Turing.jl' would appear earlier. Right now you have to wait ~ 3 seconds before it appears, and it looks like the user is just looking at empty space.

Definitely! I'll change that.

>   Would it be possible to move the text to the left-hand side and the peaks of the Gaussians to the right-hand side?

That would be a great change, too! I think it'd work particularly well given that "Turing.jl" will be appearing first, and the Gaussians then continue "unfolding" on the right of the text.

> * I like the libraries page a lot, but I don't think it deserves to be put front and centre together with tutorials and the link to the github organisation. The average user wouldn't really be navigating there very often imo.
>   * In its place, perhaps we could have a link to #turing on Julia Slack (https://julialang.slack.com/archives/CCYDC34A0) and/or the probabilistic programming forum on Discourse (https://discourse.julialang.org/c/domain/probprog/48). These are places where people can get in touch with us / other users. I think that would be both more welcoming and pertinent

Great! I believe I just looked at some other docs and what they put there, but most of the content was really just intended to be placeholders. Slack and / or discourse sound like great choices.

>   * I think the libraries page would be a good fit for the 'learn more' section at the bottom of the page

I'll move it there!

> * Is there a reason why the tutorial button is a different colour from the others? If there isn't a very important reason, I would prefer to have it be the same as UI consistency is really important.

No, no particular reason. I was probably taking inspiration from [other docs](https://geo.makie.org/v0.7.8/) and [pages](https://reactnative.dev/) that were linked as potential inspiration by @shravanngoswamii in #101.
The only reason I could see from a UX perspective is that we might want new users to check out a specific resource first, and therefore highlight it. If such a prioritised resource does not exist, I agree that they should be the same colour.

penelopeysm commented:
> If I had to argue for it from a UX perspective, I'd probably say that we might want to nudge new users to check out a specific resource first, and therefore highlight that button.

Sure, that makes sense :)

I wonder if you think this layout might be better? (it's just an Inkscape mockup) Happy to stick with the original too. It could just be that my brain is addicted to symmetry 😅 

<img width="482" alt="Screenshot 2024-11-24 at 19 28 23" src="https://github.com/user-attachments/assets/2600d80b-f440-4774-b50f-508ceb8f2e5c">

simonsteiger commented:
> I wonder if you think this layout might be better? (it's just an Inkscape mockup) Happy to stick with the original too. It could just be that my brain is addicted to symmetry 😅

I can relate to that! My attempts so far haven't worked well. Flexbox (or at least my knowledge of it) does not seem to easily accommodate this layout. Grid might be better for that, I guess?

Apart from the obvious centering issue in the current version, I still feel like the extra row creates too much space. I'll keep trying though!

<img width="1203" alt="Screenshot 2024-11-24 at 21 51 31" src="https://github.com/user-attachments/assets/dca23665-a8da-487d-8086-0d718e9b5084">
<img width="1203" alt="image" src="https://github.com/user-attachments/assets/ed0fba43-e750-4ead-a0a7-821500156746">

penelopeysm commented:
Eek, I can see what you mean with the alignment, and I guess it's because 'Slack' and 'GitHub' have different widths? That's super annoying haha. You could make it work with grid but I'm not sure if that makes the responsive design more painful too. I'm totally fine with sticking to a single line!

As for the SVG, I dug a little around the organisation and I think this is the script that's used to generate them:

https://github.com/TuringLang/turing-logo-graphics/blob/master/src/logo.jl

I ran it and it tells me that

- purple has mean 0 and stdev 1
- brown has mean 2.5 and stdev 1/sqrt(2)
- green has mean 2 and stdev 1/sqrt(3)

simonsteiger commented:
I guess we could remedy the "different lengths" problem of GitHub and Slack by setting a minimum width to all buttons, but that doesn't feel very elegant, and a short word like "Slack" might then look lost in a very wide button.

Thank you for pointing me to the script with the Gaussians! 
It seems that both Plots and Makie export these curves very large arrays of points compared to the compact bezier curve representation that @Govindggupta created. Apart from the length of the arrays, it seems that these points cannot be animated as lines. In the end, I resorted to asking Claude if it could mirror the curves – and yes, it could!

I'll work on the footer next and continue iterating a bit with colours, in particular for the links since I agree that those aren't great. @shravanngoswamii also mentioned the dark mode in general – let me know what you think should be adapted.

I would really like the header to be centered, too. I have some ideas I'd like to try, but I feel that quarto's defaults are getting in my way here, and I'm not super keen on wrangling my way past those. As @penelopeysm pointed out, a framework like Astro would be a great way to create (and reuse) components like a custom header.

simonsteiger commented:
I'd be happy to help brainstorm ideas for the "logo" on the landing page but I don't know enough about, e.g., Javis.jl to help with creating the animation in the near future.

Might it be an option to make the new animation a target for another PR?

simonsteiger commented:
In the last changes, I have

- shortened the footer text as suggested @shravanngoswamii in a comment in #101 
- replaced links to subpages for inofficial tutorials and talks with new subsections on the landing page (on the workshop section: if the Turing workshop refers to Jose Storopolis work, I think the it should go under inofficial tutorials, too)
- added a placeholders for a table on MCMC algorithms and for the Bayesian workflow diagram as listed in #76 

I can add the talks, the workflow diagram, the MCMC table, and more inofficial tutorials if you post the info.

If someone would briefly outline what information should be included in the texts for each of the subsections (e.g., next to the model examples), I can do my best to turn those points into proper texts (not a native speaker here though!).

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,224 @@
+    <div class="panel-title pb-1">
+      Intuitive
+    </div>
+    Turing models are easy to write and communicate — syntax is close to mathematical notations. 
```
```suggestion
    Turing models are easy to write and communicate — their syntax closely resembles mathematical notation. 
```

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,224 @@
+<div class="d-flex flex-row flex-wrap panel-wrapper gap-4 section-end-space">
+  <div class="panel">
+    <div class="panel-title pb-1">
+      Intuitive
```
```suggestion
      Expressive
```

I think this is a more precise word than intuitive, especially in the context of a programming language or DSL – the point is that it makes it easy for a user to put what's in their mind into code (i.e. express their ideas)

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,224 @@
+  </div>
+  <div class="panel">
+    <div class="panel-title pb-1">
+      Modular & composable
+    </div>
+    Turing is modular, written entirely in Julia, and is interoperable with the powerful Julia ecosystem.
```
```suggestion
      Composable
    </div>
    Turing is written entirely in Julia, and is interoperable with its powerful ecosystem.
```

The word 'modular' isn't explained. The modular aspect of Turing is not meaningful for the average basic user anyway, so I think it's easier to leave it out rather than to try explaining it.

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,224 @@
 
-### Modular and composable 
-Turing is modular, written entirely in Julia, and is interoperable with the powerful Julia ecosystem. 
+    return y
```
```suggestion
```

When I got into Turing I personally found return values to be rather confusing. Basic sampling doesn't make use of the return value (you could `return nothing`, or just omit the return statement, and `sample(coinflip; 100) | (y=y,)` still works exactly the same way).

I would prefer to remove it because including it implies that it is needed for the example to work

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,224 @@
+      Ecosystem
+    </div>
+    <p class="quarto-grid-item">
+      Learn about all the packages that are part of The Turing Organization.
```
```suggestion
      Learn about all the packages that are part of the TuringLang organization.
```

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,224 @@
+      Documentation
+    </div>
+    <p class="quarto-grid-item">
+      Familiarize yourself with the Turing.jl API.
```
```suggestion
      Familiarize yourself with the mechanics of Turing.jl.
```

API docs aren't on the docs website, they're on the individual repos (e.g. https://turinglang.org/Turing.jl). (Though this does makes me think we should at least mention this on the docs website)

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,224 @@
+      Official tutorials
+    </div>
+    <p class="quarto-grid-item">
+      Learn to tackle specific problems with Turing.jl.
```
```suggestion
      Learn to tackle specific modelling problems with Turing.jl.
```

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,224 @@
+  </ul>
+</div>
+<div class="flex-grow-1">
+  <h3>Inofficial tutorials</h3>
+  <p>Placeholder for intro text on list of inofficial tutorials</p>
```
```suggestion
  <h3>Other resources</h3>
  <p>This list contains a few other written resources for learning Turing. If you have written something and would like to share it with the community, please <a href="https://github.com/TuringLang/docs/issues/new">get in touch</a>!</p>
```

penelopeysm replied:
Not sure if the docs repo is the best place to raise an issue. I figured it would either be the docs repo or this repo, but this one is quite low-profile so it might not get noticed very soon.

penelopeysm commented on index.qmd:
```diff
@@ -1,55 +1,224 @@
+  <p>Placeholder for intro text on list of inofficial tutorials</p>
+  <ul>
+    <li><a href="https://github.com/StatisticalRethinkingJulia/SR2TuringPluto.jl">Statistical Rethinking</a>
+    <li><a href="storopoli.github.io/Turing-Workshop/">Jose Storopoli's Turing workshop</a>
+    <li><a href="https://storopoli.io/Bayesian-Julia/">Jose Storopoli's blog on Turing and Bayesian Statistics</a>
```
```suggestion
    <li><a href="https://storopoli.github.io/Turing-Workshop/">Jose Storopoli's Turing workshop (outdated in parts)</a>
    <li><a href="https://storopoli.github.io/Bayesian-Julia/">Jose Storopoli's blog on Turing and Bayesian Statistics</a>
```

The links don't work. Also, the workshop is indeed outdated in parts 😄

penelopeysm reviewed (changes requested):
Just going through the text itself. I think some of them predate this PR, but this is as good a time as any to change things :)

penelopeysm commented:
@willtebbutt Could we trouble you to write the text for the GP model? (For your convenience, the preview for this PR is at https://turinglang.org/pr-previews/105/)

willtebbutt commented:
Sorry -- this slipped through the cracks somehow. What exactly is it that you would like me to do @penelopeysm ?

penelopeysm commented:
@willtebbutt There's this placeholder text "Goodbye, World in Turing" next to a GP model. I thought you'd be the right person to ask about filling that in with real text 😄

willtebbutt commented on index.qmd:
````diff
@@ -1,55 +1,223 @@
+  <div class="fs-4 fw-bold pb-1">
+  Goodbye, World in Turing
+  </div>
+  Some text about how easy it is to interface with external packages like AbstractGPs. Learn more about modelling [Gaussian Processes](https://turinglang.org/docs/tutorials/15-gaussian-processes/) with Turing.jl.
+</div>
+<div class="example-code">
+```julia
+@model function putting_model(d, n; jitter=1e-4)
+    v ~ Gamma(2, 1)
+    l ~ Gamma(4, 1)
+    f = GP(v * with_lengthscale(SEKernel(), l))
+    f_latent ~ f(d, jitter)
+    binomials = Binomial.(n, logistic.(f_latent))
+    y ~ product_distribution(binomials)
+    return (fx=f(d, jitter), f_latent=f_latent, y=y)
+end
+```
````
If I'm being totally honest, I'd be inclined to remove this example, and just have the link to the GPs page. Having the code here, and not making it something which gets run every time that the docs are built, is just asking for it to become outdated at some point.

In terms of the text, I'd really just say something along the lines of `anything which behaves like a Distributions.jl distribution plays nicely with Turing.jl, e.g. [Gaussian processes](https://turinglang.org/docs/tutorials/15-gaussian-processes/)`

simonsteiger replied:
That makes sense. We could just add the text as a new paragraph under "Hello, World in Turing" and use the space that the GP example took up for something else – maybe presenting the samplers? I recall that the old [turing.ml](turing.ml) website had something similar.

<img width="1407" alt="image" src="https://github.com/user-attachments/assets/b6c29c62-8785-43a5-aa72-a792f9920f7f" />

simonsteiger commented:
It seems that quarto 1.6.39 treats the first raw html `h1` as the page title and therefore wraps it in `quarto-title-block`. That caused the CSS formatting I had applied to said `h1` to be dropped.

For now, I have simply replaced `h1` by `div`. This means "Turing.jl" doesn't have the same font weight at the moment, but we can of course get that back with a custom class.
